src/docs/rfcs/018-declarative-vdu.md

```diff
@@ -0,0 +1,2249 @@
+equality constraint. i.e. `{ "a": { "$data": "b" } }` means `{ "a": { "$eq": {
+"$data": "b" } } }`, otherwise this would allow operator injection.
+
+
+### `$cat`
+
+The `$cat` operator takes an array whose elememts are either literal strings, or
+`$data` references, and produces the result of concatenating them after
+resolving the `$data` references. For example:
+
+    {
+      "_id": { "$cat": ["org.couchdb.user:", { "$data": "name" }] }
+    }
+
+This means that the `_id` field must equal the `name` field with
+`org.couchdb.user:` as a prefix.
+
+`$cat` has the same restrictions on its use as the `$data` field, i.e. it may
+only appear where a literal value is expected, and must not appear anywhere a
+sub-selector expression is expected. When it appears as the matcher for an
+object property, it should be understood as a strict equality constraint, i.e.
+the above expression means the same as:
+
+    {
+      "_id": {
+        "$eq": {
+          "$cat": ["org.couchdb.user:", { "$data": "name" }]
+        }
+      }
+    }
+
+
+### `$ref`
+
+Declarative VDUs can make use of reusable expressions when defining their logic.
+Definitions of such expressions are placed in the `defs` field of the design
+document, and then referenced within `validate_doc_update` expressions via the
+`$ref` operator.
+
+For example, to define an expression for matching even numbers, this structure
+is placed in the root of the design document:
+
+    "defs": {
+      "even-number": {
+        "$type": "number",
+        "$mod": [2, 0]
+      }
+    }
+
+Then, a field can be validated as an even number using this syntax:
+
+    "some_field": { "$ref": "defs.even-number" }
+
+The `$ref` operator takes a path with the same syntax as `$data`, but it is
+resolved relative to the root of the design document.
+
+The effect of using `$ref` and other operators in the same selector should be
+the same as if the `$ref` expression where combined with the rest of the
+selector via `$and`. For example, `{ "$ref": "devs.even-number", "$gt": 20 }`
+should have the same effect as:
+
+    {
+      "$and": [
+        { "$ref": "defs.even-number" },
+        { "$gt": 20 }
+      ]
+    }
+
+If the implementation wishes to inline `$ref` expressions before evaluation, it
+should first translate any uses of `$ref` into the above form to avoid key
+collisions when merging objects.
+
+This operator may be used to define recursive structures. For example, if we had
+a way of representing a tree of HTML nodes, where each node has a tag name, a
+set of attributes, and a list of child nodes, we could validate it using this
+definition:
+
+    "defs": {
+      "html-tree": {
+        "tagName": { "$type": "string" },
+        "attributes": { "$type": "object" },
+        "children": {
+          "$type": "array",
+          "$allMatch": { "$ref": "defs.html-tree" }
+        }
+      }
+    }
+
+**To be decided**: we may wish to prohibit such recursion, as doing so would let
```

Comment:
While the ref idea is neat, maybe it's something to be added later, in phase 2, just to avoid adding a lot of initial complexity. I could see a few levels of data and refs becoming tricky for users to follow and understand and like with the recursion case, it would be a bit tricky to implement.